### PR TITLE
feat(amazonqFeatureDev): add partial code acceptance

### PR DESCRIPTION
**Pull request by sannicm**
## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)

## Description

Mynah UI already supports an option to handle file clicking. This is being leveraged to add a partial code acceptance feautre to the /dev chat. That is, the user will be able to click through the file changes generated from code generation step to reject them or not. 

## Checklist
- [x] My code follows the code style of this project
- [n/a] I have added tests to cover my changes
   -  will be added in a follow up
- [n/a] A short description of the change has been added to the **[CHANGELOG](https://github.com/aws/aws-toolkit-jetbrains/blob/master/CONTRIBUTING.md#contributing-via-pull-requests)** if the change is customer-facing in the IDE.
- [n/a] I have added metrics for my changes (if required)
 
## License
I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: plugins/amazonq/chat/jetbrains-community/src/software/aws/toolkits/jetbrains/services/amazonqFeatureDev/messages/FeatureDevMessagePublisherExtensions.kt [hunk 1]
```diff
@@ -56,6 +57,15 @@ suspend fun MessagePublisher.sendSystemPrompt(
     )
 }
 
+suspend fun MessagePublisher.updateFileComponent(tabId: String, filePaths: List<NewFileZipInfo>, deletedFiles: List<DeletedFileInfo>) {
+    val fileComponentMessage = FileComponent(
```
Comment (kumsmrit) [inline plugins/amazonq/chat/jetbrains-community/src/software/aws/toolkits/jetbrains/services/amazonqFeatureDev/messages/FeatureDevMessagePublisherExtensions.kt:61]: What is this being used for, how will the customer experience be?

File: plugins/amazonq/chat/jetbrains-community/src/software/aws/toolkits/jetbrains/services/amazonqFeatureDev/session/Session.kt [hunk 2]
```diff
@@ -90,22 +91,35 @@ class Session(val tabID: String, val project: Project) {
...
-    fun insertChanges(filePaths: List<NewFileZipInfo>, deletedFiles: List<String>, references: List<CodeReference>) {
+    fun insertChanges(filePaths: List<NewFileZipInfo>, deletedFiles: List<DeletedFileInfo>, references: List<CodeReference>) {
         val projectRootPath = context.projectRoot.toNioPath()
 
-        filePaths.forEach {
-            val filePath = projectRootPath.resolve(it.zipFilePath)
-            filePath.parent.createDirectories() // Create directories if needed
-            filePath.writeBytes(it.fileContent.toByteArray(Charsets.UTF_8))
-        }
+        filePaths
```
Comment (dianaabv16) [inline plugins/amazonq/chat/jetbrains-community/src/software/aws/toolkits/jetbrains/services/amazonqFeatureDev/session/Session.kt:109]: if possible can we follow up on adding test for insertChanges. 
Same way we have here [link](https://github.com/aws/aws-toolkit-vscode/pull/4594)

File: plugins/amazonq/chat/jetbrains-community/src/software/aws/toolkits/jetbrains/services/amazonqFeatureDev/controller/FeatureDevController.kt [hunk 3]
```diff
@@ -268,7 +304,7 @@ class FeatureDevController(
             }
             AmazonqTelemetry.isAcceptedCodeChanges(
                 project = null,
-                amazonqNumberOfFilesAccepted = (filePaths.size + deletedFiles.size) * 1.0,
+                amazonqNumberOfFilesAccepted = (filePaths.filterNot { it.rejected }.size + deletedFiles.filterNot { it.rejected }.size) * 1.0,
```
Comment (dianaabv16) [inline plugins/amazonq/chat/jetbrains-community/src/software/aws/toolkits/jetbrains/services/amazonqFeatureDev/controller/FeatureDevController.kt:307]: in vscode we calculate differently [link](https://github.com/aws/aws-toolkit-vscode/pull/4523/files#diff-3126666ce5709a5f19fcabd5c734242bf55ac7ba9dff730b5393d68911693012R408)

can you add a task to a board to fix vscode metrics please?

File: plugins/amazonq/chat/jetbrains-community/src/software/aws/toolkits/jetbrains/services/amazonqFeatureDev/controller/FeatureDevController.kt [hunk 4]
```diff
@@ -207,6 +208,41 @@ class FeatureDevController(
...
+            var deletedFiles: List<DeletedFileInfo> = emptyList()
+
+            when (val state = session.sessionState) {
+                is PrepareCodeGenerationState -> {
+                    filePaths = state.filePaths
+                    deletedFiles = state.deletedFiles
+                }
+            }
+            filePaths.find { it.zipFilePath == fileToUpdate }?.let { it.rejected = !it.rejected }
+            deletedFiles.find { it.zipFilePath == fileToUpdate }?.let { it.rejected = !it.rejected }
```
Comment (vishaalmehrishi) [inline plugins/amazonq/chat/jetbrains-community/src/software/aws/toolkits/jetbrains/services/amazonqFeatureDev/controller/FeatureDevController.kt:227]: The find condition is exactly the same for both

Comment (vishaalmehrishi) [reply plugins/amazonq/chat/jetbrains-community/src/software/aws/toolkits/jetbrains/services/amazonqFeatureDev/controller/FeatureDevController.kt:227]: Nvm, now I understand why. Please ignore this comment.

File: plugins/amazonq/chat/jetbrains-community/src/software/aws/toolkits/jetbrains/services/amazonqFeatureDev/controller/FeatureDevController.kt [hunk 5]
```diff
@@ -207,6 +208,41 @@ class FeatureDevController(
...
+                tabId = message.tabId,
+                filePaths = filePaths,
+                deletedFiles = deletedFiles
+            )
+        } catch (err: Exception) {
+            val errorMessage = createUserFacingErrorMessage("Failed to process diff tree file clicked: ${err.message}")
+            messenger.sendError(
+                tabId = message.tabId,
+                errMessage = errorMessage ?: message("amazonqFeatureDev.exception.insert_code_failed"),
+                retries = retriesRemaining(session),
```
[the remark this answers is not in the document]
Comment (sannicm) [reply plugins/amazonq/chat/jetbrains-community/src/software/aws/toolkits/jetbrains/services/amazonqFeatureDev/controller/FeatureDevController.kt:240]: I don't think so, I'll remove.

File: plugins/amazonq/chat/jetbrains-community/src/software/aws/toolkits/jetbrains/services/amazonqFeatureDev/controller/FeatureDevController.kt (hunk 3 above)
Comment (sannicm) [reply plugins/amazonq/chat/jetbrains-community/src/software/aws/toolkits/jetbrains/services/amazonqFeatureDev/controller/FeatureDevController.kt:307]: Yes, done: https://taskei.amazon.dev/tasks/D127722149

File: plugins/amazonq/mynah-ui/src/mynah-ui/ui/connector.ts [hunk 6]
```diff
@@ -41,13 +42,15 @@ export interface ConnectorProps {
     onCWCOnboardingPageInteractionMessage: (message: ChatItem) => string | undefined
     onError: (tabID: string, message: string, title: string) => void
     onWarning: (tabID: string, message: string, title: string) => void
+    onFileComponentUpdate: (tabID: string, filePaths: DiffTreeFileInfo[], deletedFiles: DiffTreeFileInfo[]) => void
```
[the remark this answers is not in the document]
Comment (sannicm) [reply plugins/amazonq/mynah-ui/src/mynah-ui/ui/connector.ts:45]: Not really, in fact i think I need to remove it from here. I'll double check.

[the remark this answers is not in the document]
Comment (sannicm) [reply plugins/amazonq/mynah-ui/src/mynah-ui/ui/connector.ts:45]: In summary:

original codeResult message -> onFileClick -> processfileClick -> updatefileComponent -> updated codeResultMessage

File: plugins/amazonq/chat/jetbrains-community/src/software/aws/toolkits/jetbrains/services/amazonqFeatureDev/messages/FeatureDevMessagePublisherExtensions.kt (hunk 1 above)
Comment (sannicm) [reply plugins/amazonq/chat/jetbrains-community/src/software/aws/toolkits/jetbrains/services/amazonqFeatureDev/messages/FeatureDevMessagePublisherExtensions.kt:61]: This is answered in your other question, but to also give more context, the experience is that the customer clicks on rejection/revert rejection and the file line in the file tree is higlighted with red and has a X icon that gives feedback tot he customer that the file is being rejected. This is the message to tell mynahUI to update the UI component.

File: plugins/amazonq/mynah-ui/src/mynah-ui/ui/connector.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (sannicm) [reply plugins/amazonq/mynah-ui/src/mynah-ui/ui/connector.ts:45]: Just checked again, it is being in fact handled specifically in main.ts https://github.com/aws/aws-toolkit-jetbrains/pull/4244/files#diff-deb4fc333346795314beae818407e5aa42d7cb6b7e4714862e591f8814fccfeaR268

File: plugins/amazonq/chat/jetbrains-community/src/software/aws/toolkits/jetbrains/services/amazonqFeatureDev/controller/FeatureDevController.kt [hunk 7]
```diff
@@ -207,6 +208,31 @@ class FeatureDevController(
         }
     }
 
+    override suspend fun processFileClicked(message: IncomingFeatureDevMessage.FileClicked) {
+        val fileToUpdate = message.filePath
+        val session = getSessionInfo(message.tabId)
+
+        var filePaths: List<NewFileZipInfo> = emptyList()
+        var deletedFiles: List<DeletedFileInfo> = emptyList()
```
Comment (rli) [inline plugins/amazonq/chat/jetbrains-community/src/software/aws/toolkits/jetbrains/services/amazonqFeatureDev/controller/FeatureDevController.kt:216]: var can be val

**Review by dianaabv16 — APPROVED** [review]

**Review by kumsmrit — APPROVED** [review]

**Review by rli — APPROVED** [review]

**Review by dogusata — APPROVED** [review]

**Pull request merged**